### PR TITLE
[WIP] Variable length trials fix

### PR DESCRIPTION
**Pull request by sliwy**
Just two small changes to fix cropped decoding to work with variable length trials. Tests does not work locally in my case (no clue why), so let's check here. Trying to solve #222

### CODE REVIEW

File: braindecode/classifier.py [hunk 1]
```diff
@@ -221,7 +221,7 @@ def predict_proba(self, X):
 
         """
         y_pred = super().predict_proba(X)
-        if self.cropped:
+        if self.cropped and len(y_pred.shape) == 3:
```
Comment (robintibor) [inline braindecode/classifier.py:224]: Please add some comment here why this happens! And what we expect etc....

File: braindecode/training/scoring.py [hunk 2]
```diff
@@ -177,7 +177,9 @@ def on_epoch_end(self, net, dataset_train, dataset_valid, **kwargs):
                 pred_results['i_window_stops'])
             # trial preds is a list
             # each item is an 2d array classes x time
-            y_preds_per_trial = np.array(trial_preds)
+            y_preds_per_trial = np.array(
```
Comment (robintibor) [inline braindecode/training/scoring.py:180]: Also here some comments why we are averaging here would be great I think

**Review by robintibor — CHANGES_REQUESTED** [review]

**Comment by robintibor** [comment]
Great I think this can be merged, but would be great if you can add some comments so we know what we were doing there and why

File: braindecode/classifier.py (hunk 1 above)
Comment (robintibor) [reply braindecode/classifier.py:224]:
```suggestion
        if self.cropped and len(y_pred.shape) == 3:
```

```suggestion
# Normally, we have to average the predictions across crops/timesteps
# to get one prediction per time window/trial
# In the cropped scoring , we may have averaged these predictions 
# already, hence the check if len(y_pred.shape) == 3:
        if self.cropped and len(y_pred.shape) == 3:
```

File: braindecode/training/scoring.py [hunk 3]
```diff
@@ -177,7 +177,9 @@ def on_epoch_end(self, net, dataset_train, dataset_valid, **kwargs):
...
+                [np.mean(p, axis=1) for p in trial_preds]
+            )
```
Comment (robintibor) [inline braindecode/training/scoring.py:182]:
```suggestion
            # Average across the timesteps of each trial so we have per-trial predictions already,
            # these will be just passed through the forward method of the classifier/regressor
            # to the skorch scoring function
            y_preds_per_trial = np.array(
                [np.mean(p, axis=1) for p in trial_preds]
            )
```

Comment (robintibor) [reply braindecode/training/scoring.py:182]: Do we also need to adapt the forward function of the regressor @sliwy ?

**Review by robintibor — CHANGES_REQUESTED** [review]

File: braindecode/classifier.py (hunk 1 above)
Comment (sliwy) [reply braindecode/classifier.py:224]: this is what I left for later (together with the comments to do) if the changes from this PR works with your example. @robintibor have you checked if ti works with your example with variable trial length?

Comment (robintibor) [reply braindecode/classifier.py:224]: I checked it in my mind and it seemed to work! Now also checked in reality and also worked! :)

Comment (robintibor) [reply braindecode/classifier.py:224]: so do you have capacity to finish it? @sliwy  Otherwise I can have a try :)

Comment (sliwy) [reply braindecode/classifier.py:224]: @robintibor I can take care of it today or tomorrow evening. If you need it faster feel free to finish that. :)

Comment (robintibor) [reply braindecode/classifier.py:224]: sounds good! should be quick enough! :)

**Comment by sliwy** [comment]
I added comments you asked for, I also modified EEGRegressor to work the same way as EEGClassifier. 

I still have some doubts whether the `predict_proba` should work with variable length trials (for scoring it should be ok). In particular, I don't know what will be the output of skorch `predict_proba` method when provided with observation of different shape. I've never done this before. Maybe we should create a test for that? It would need a custom DataLoader?  @robintibor  As you've already tested it on your side maybe you have a minimal working example for variable length trials decoding that we can use a starting point for the test?

**Comment by robintibor** [comment]
I think my example is too complex, also requires loading TUH etc., but I can also see whether I can distill something out of it, if you have time before to make a test, also great, otherwise I will see

**Comment by robintibor** [comment]
ok since my colleague needs it quite urgently now, merging this and we can do test in separate PR potentially.

**Review by robintibor — APPROVED** [review]

**Pull request merged**